presto-main/src/main/java/io/prestosql/operator/NestedLoopJoinOperator.java

```diff
@@ -235,59 +242,132 @@ public void close()
         afterClose.run();
     }
 
+    @VisibleForTesting
+    static NestedLoopOutputIterator createNestedLoopOutputIterator(Page probePage, Page buildPage, int[] probeChannels, int[] buildChannels)
+    {
+        if (probeChannels.length == 0 && buildChannels.length == 0) {
+            int probePositions = probePage.getPositionCount();
+            int buildPositions = buildPage.getPositionCount();
+            try {
+                // positionCount is an int. Make sure the product can still fit in an int.
+                int outputPositions = multiplyExact(probePositions, buildPositions);
+                return new PageRepeatingIterator(new Page(outputPositions), 1);
+            }
+            catch (ArithmeticException ignored) {
+            }
+            // Repeat larger position count a smaller position count number of times
+            Page outputPage = new Page(max(probePositions, buildPositions));
+            return new PageRepeatingIterator(outputPage, min(probePositions, buildPositions));
+        }
+        else if (probeChannels.length == 0 && probePage.getPositionCount() <= buildPage.getPositionCount()) {
+            return new PageRepeatingIterator(buildPage.getColumns(buildChannels), probePage.getPositionCount());
+        }
+        else if (buildChannels.length == 0 && buildPage.getPositionCount() <= probePage.getPositionCount()) {
+            return new PageRepeatingIterator(probePage.getColumns(probeChannels), buildPage.getPositionCount());
+        }
+        else {
+            return new NestedLoopPageBuilder(probePage, buildPage, probeChannels, buildChannels);
+        }
+    }
+
+    // bi-morphic parent class to avoid invoking iterator interface methods via invokeinterface on the hot path
```

Comment:
The performance issue isn't the invokeinterface vs invokevirtual really, both would be inlined with direct dispatch *so long as the call site is actually bimorphic*. The abstract parent class and this comment are here to explain that it's important that the implementation stay bimorphic and doesn't become megamorphic by in some future change which might otherwise be very easy to introduce accidentally.

Reply:
I've clarified the comment to better describe the original intention behind it and made the parent class not implement `Iterator<Page` to the same end. Let me know if that's enough to satisfy the concern here or if you'd like another flavor instead.